File: atcoder/312/B_TaK_Code.cpp
#include <iostream>
#include <vector>
using namespace std;

bool isBlack(char cell) {
    return cell == '#';
}

bool isWhite(char cell) {
    return cell == '.';
}

bool isValidBlock(const vector<string>& grid, int row, int col) {
    // Check if the block is a 3x3 region of black cells
    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            if (!isBlack(grid[row + i][col + j])) {
                return false;
            }
        }
    }

    // Check if there are no black cells adjacent to the block
    int adjacent_r[] = {-1, -1, -1, 0, 0, 1, 1, 1};
    int adjacent_c[] = {-1, 0, 1, -1, 1, -1, 0, 1};

    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            for (int k = 0; k < 8; ++k) {
                int r = row + i + adjacent_r[k];
                int c = col + j + adjacent_c[k];
                if (r >= 0 && r < grid.size() && c >= 0 && c < grid[0].size() && isBlack(grid[r][c])) {
                    return false;
                }
            }
        }
    }

    return true;
}


int main() {
    int N, M;
    cin >> N >> M;
    vector<string> grid(N);
    for (int i = 0; i < N; ++i) {
        cin >> grid[i];
    }

        
    for (int i = 0; i <= N - 9; ++i) {
        for (int j = 0; j <= M - 9; ++j) {
            if (isBlack(grid[i][j]) && isValidBlock(grid, i, j) && isValidBlock(grid, i + 6, j + 6)) {
                cout<<i+1<<" "<<j+1<<endl;
            }
        }
    }


   

    return 0;
}

File: atcoder/312/B_TaK_Code.py
def is_black(cell):
    return cell == '#'

def is_white(cell):
    return cell == '.'

def is_tak_code_region(grid, row, col):
    # Check condition 1: All 18 cells in the top-left and bottom-right three-by-three regions are black.
    if not all(all(is_black(grid[row + i][col + j]) for j in range(3)) for i in range(3)):
        return False

    if not all(all(is_black(grid[row + i][col + j]) for j in range(6, 9)) for i in range(6, 9)):
        return False

    # Check condition 2: All 14 cells adjacent to the top-left or bottom-right three-by-three region are white.
    adjacent_cells = [
        (row - 1, col - 1), (row - 1, col), (row - 1, col + 1),
        (row, col - 1), (row, col + 1),
        (row + 1, col - 1), (row + 1, col), (row + 1, col + 1)
    ]
    for r, c in adjacent_cells:
        if not is_white(grid[r][c]):
            return False

    return True

def find_tak_codes(N, M, grid):
    tak_codes = []
    for i in range(N - 8):
        for j in range(M - 8):
            if is_tak_code_region(grid, i, j):
                tak_codes.append((i + 1, j + 1))  # Adjusting to 1-based indexing

    return tak_codes

# Read input
N, M = map(int, input().split())
grid = [input() for _ in range(N)]

# Find TaK Codes
result = find_tak_codes(N, M, grid)

# Output the result
for i, j in result:
    print(i, j)
